# Page Visits Funnel

Cool T-Shirts Inc. has asked you to analyze data on visits to their website. Your job is to build a funnel, which is a description of how many people continue to the next step of a multi-step process.

In [5]:
import pandas as pd

In [6]:
visits = pd.read_csv('visits.csv')
cart = pd.read_csv('cart.csv')
checkout = pd.read_csv('checkout.csv')
purchase = pd.read_csv('purchase.csv')

In [7]:
visits.head()

user_id           visit_time
0  943647ef-3682-4750-a2e1-918ba6f16188  2017-04-07 15:14:00
1  0c3a3dd0-fb64-4eac-bf84-ba069ce409f2  2017-01-26 14:24:00
2  6e0b2d60-4027-4d9a-babd-0e7d40859fb1  2017-08-20 08:23:00
3  6879527e-c5a6-4d14-b2da-50b85212b0ab  2017-11-04 18:15:00
4  a84327ff-5daa-4ba1-b789-d5b4caf81e96  2017-02-27 11:25:00

In [8]:
visits.duplicated().sum()

0

In [9]:
len(visits)

2000

In [10]:
cart.head()

user_id            cart_time
0  2be90e7c-9cca-44e0-bcc5-124b945ff168  2017-11-07 20:45:00
1  4397f73f-1da3-4ab3-91af-762792e25973  2017-05-27 01:35:00
2  a9db3d4b-0a0a-4398-a55a-ebb2c7adf663  2017-03-04 10:38:00
3  b594862a-36c5-47d5-b818-6e9512b939b3  2017-09-27 08:22:00
4  a68a16e2-94f0-4ce8-8ce3-784af0bbb974  2017-07-26 15:48:00

In [11]:
cart.duplicated().sum()

0

In [12]:
len(cart)

348

In [13]:
checkout.head()

user_id        checkout_time
0  d33bdc47-4afa-45bc-b4e4-dbe948e34c0d  2017-06-25 09:29:00
1  4ac186f0-9954-4fea-8a27-c081e428e34e  2017-04-07 20:11:00
2  3c9c78a7-124a-4b77-8d2e-e1926e011e7d  2017-07-13 11:38:00
3  89fe330a-8966-4756-8f7c-3bdbcd47279a  2017-04-20 16:15:00
4  3ccdaf69-2d30-40de-b083-51372881aedd  2017-01-08 20:52:00

In [14]:
len(checkout)

360

In [15]:
checkout.duplicated().sum()

5

In [16]:
checkout = checkout.drop_duplicates().reset_index(drop=True)
checkout.duplicated().sum()

0

In [17]:
len(checkout)

355

In [18]:
purchase.head()

user_id        purchase_time
0  4b44ace4-2721-47a0-b24b-15fbfa2abf85  2017-05-11 04:25:00
1  02e684ae-a448-408f-a9ff-dcb4a5c99aac  2017-09-05 08:45:00
2  4b4bc391-749e-4b90-ab8f-4f6e3c84d6dc  2017-11-20 20:49:00
3  a5dbb25f-3c36-4103-9030-9f7c6241cd8d  2017-01-22 15:18:00
4  46a3186d-7f5a-4ab9-87af-84d05bfd4867  2017-06-11 11:32:00

In [19]:
len(purchase)

252

In [20]:
purchase.duplicated().sum()

4

In [21]:
purchase = purchase.drop_duplicates().reset_index(drop=True)
purchase.duplicated().sum()

0

In [22]:
len(purchase)

248

Combine visits and cart using a left merge.

In [23]:
visits_cart = pd.merge(visits, cart, how='left')
visits_cart

user_id           visit_time  \
0     943647ef-3682-4750-a2e1-918ba6f16188  2017-04-07 15:14:00   
1     0c3a3dd0-fb64-4eac-bf84-ba069ce409f2  2017-01-26 14:24:00   
2     6e0b2d60-4027-4d9a-babd-0e7d40859fb1  2017-08-20 08:23:00   
3     6879527e-c5a6-4d14-b2da-50b85212b0ab  2017-11-04 18:15:00   
4     a84327ff-5daa-4ba1-b789-d5b4caf81e96  2017-02-27 11:25:00   
...                                    ...                  ...   
1995  33913ac2-03da-45ae-8fc3-fea39df827c6  2017-03-25 03:29:00   
1996  4f850132-b99d-4623-80e6-6e61d003577e  2017-01-08 09:57:00   
1997  f0830b9b-1f5c-4e74-b63d-3f847cc6ce70  2017-09-07 12:56:00   
1998  b01bffa7-63ba-4cd3-9d93-eb1477c23831  2017-07-20 04:37:00   
1999  0336ca81-8d68-443f-9248-ac0b8ad147d5  2017-11-15 10:11:00   

                cart_time  
0                     NaN  
1     2017-01-26 14:44:00  
2     2017-08-20 08:31:00  
3                     NaN  
4                     NaN  
...                   ...  
1995                  NaN  
1996                  NaN  
1997                  NaN  
1998                  NaN  
1999                  NaN  

[2000 rows x 3 columns]

How many of the timestamps are null for the column cart_time?

What do these null rows mean?

In [24]:
null_visits_cart = len(visits_cart[visits_cart.cart_time.isnull()])
null_visits_cart

1652

What percent of users who visited Cool T-Shirts Inc. ended up not placing a t-shirt in their cart?

In [25]:
percent_visits_cart = 100*float(null_visits_cart / (len(visits_cart)))
percent_visits_cart

82.6

Combine cart and checkout using a left merge.

In [26]:
cart_checkout = pd.merge(cart, checkout, how='left')
cart_checkout

user_id            cart_time  \
0    2be90e7c-9cca-44e0-bcc5-124b945ff168  2017-11-07 20:45:00   
1    2be90e7c-9cca-44e0-bcc5-124b945ff168  2017-11-07 20:45:00   
2    2be90e7c-9cca-44e0-bcc5-124b945ff168  2017-11-07 20:45:00   
3    4397f73f-1da3-4ab3-91af-762792e25973  2017-05-27 01:35:00   
4    a9db3d4b-0a0a-4398-a55a-ebb2c7adf663  2017-03-04 10:38:00   
..                                    ...                  ...   
472  59d2b685-383a-4fce-a325-35d985713a8b  2017-01-20 03:58:00   
473  a1203970-92ae-4fc7-843e-e1740c1be9d5  2017-11-12 23:07:00   
474  55de53ae-89f6-43ff-a26d-df284678aca1  2017-03-17 19:56:00   
475  0ea4cc68-dae4-4e35-b3e0-f0889932e1b5  2017-05-12 08:53:00   
476  0ea4cc68-dae4-4e35-b3e0-f0889932e1b5  2017-05-12 08:53:00   

           checkout_time  
0    2017-11-07 21:14:00  
1    2017-11-07 20:50:00  
2    2017-11-07 21:11:00  
3                    NaN  
4    2017-03-04 11:04:00  
..                   ...  
472                  NaN  
473                  NaN  
474  2017-03-17 20:04:00  
475  2017-05-12 08:57:00  
476  2017-05-12 09:20:00  

[477 rows x 3 columns]

In [27]:
null_cart_checkout = len(cart_checkout[cart_checkout.checkout_time.isnull()])
null_cart_checkout

122

In [28]:
percent_cart_checkout = 100*float(null_cart_checkout / (len(cart_checkout)))
percent_cart_checkout

25.57651991614256

Combine checkout and purchase using a left merge.

In [29]:
checkout_purchase = pd.merge(checkout, purchase, how='left')
checkout_purchase

user_id        checkout_time  \
0    d33bdc47-4afa-45bc-b4e4-dbe948e34c0d  2017-06-25 09:29:00   
1    d33bdc47-4afa-45bc-b4e4-dbe948e34c0d  2017-06-25 09:29:00   
2    d33bdc47-4afa-45bc-b4e4-dbe948e34c0d  2017-06-25 09:29:00   
3    4ac186f0-9954-4fea-8a27-c081e428e34e  2017-04-07 20:11:00   
4    3c9c78a7-124a-4b77-8d2e-e1926e011e7d  2017-07-13 11:38:00   
..                                    ...                  ...   
578  21dec5fa-999a-45c5-b59b-18a1ee161379  2017-03-17 12:40:00   
579  f3dd7631-3cf6-4d49-9c6e-b48eda4432b5  2017-01-04 10:20:00   
580  ba011b1d-6cc5-4147-ae8a-1722a1f5c27c  2017-10-24 08:32:00   
581  d2cb350b-2201-4290-b2e0-84a8bf0d6883  2017-08-08 16:14:00   
582  d2cb350b-2201-4290-b2e0-84a8bf0d6883  2017-08-08 16:14:00   

           purchase_time  
0    2017-06-25 09:49:00  
1    2017-06-25 09:56:00  
2    2017-06-25 09:57:00  
3                    NaN  
4                    NaN  
..                   ...  
578  2017-03-17 12:54:00  
579                  NaN  
580                  NaN  
581  2017-08-08 16:35:00  
582  2017-08-08 16:34:00  

[583 rows x 3 columns]

In [30]:
null_checkout_purchase = len(checkout_purchase[checkout_purchase.purchase_time.isnull()])
null_checkout_purchase

101

In [31]:
percent_checkout_purchase = 100*float(null_checkout_purchase / (len(checkout_purchase)))
percent_checkout_purchase

17.32418524871355

Which step of the funnel is weakest (i.e., has the highest percentage of users not completing it)?

checkout to purchase the percentage is 17.32%

other funnels :
- visits to cart the percentage is 82.6%
- cart to checkout the percenatge is 25.57%
- checkout to purchase the percentage is 17.32%

Merge all four steps of the funnel

In [32]:
all_data = visits.merge(cart, how='left').merge(checkout, how='left').merge(purchase, how='left')
all_data

user_id           visit_time  \
0     943647ef-3682-4750-a2e1-918ba6f16188  2017-04-07 15:14:00   
1     0c3a3dd0-fb64-4eac-bf84-ba069ce409f2  2017-01-26 14:24:00   
2     6e0b2d60-4027-4d9a-babd-0e7d40859fb1  2017-08-20 08:23:00   
3     6879527e-c5a6-4d14-b2da-50b85212b0ab  2017-11-04 18:15:00   
4     a84327ff-5daa-4ba1-b789-d5b4caf81e96  2017-02-27 11:25:00   
...                                    ...                  ...   
2352  33913ac2-03da-45ae-8fc3-fea39df827c6  2017-03-25 03:29:00   
2353  4f850132-b99d-4623-80e6-6e61d003577e  2017-01-08 09:57:00   
2354  f0830b9b-1f5c-4e74-b63d-3f847cc6ce70  2017-09-07 12:56:00   
2355  b01bffa7-63ba-4cd3-9d93-eb1477c23831  2017-07-20 04:37:00   
2356  0336ca81-8d68-443f-9248-ac0b8ad147d5  2017-11-15 10:11:00   

                cart_time        checkout_time        purchase_time  
0                     NaN                  NaN                  NaN  
1     2017-01-26 14:44:00  2017-01-26 14:54:00  2017-01-26 15:08:00  
2     2017-08-20 08:31:00                  NaN                  NaN  
3                     NaN                  NaN                  NaN  
4                     NaN                  NaN                  NaN  
...                   ...                  ...                  ...  
2352                  NaN                  NaN                  NaN  
2353                  NaN                  NaN                  NaN  
2354                  NaN                  NaN                  NaN  
2355                  NaN                  NaN                  NaN  
2356                  NaN                  NaN                  NaN  

[2357 rows x 5 columns]

In [33]:
all_data.duplicated().sum()

0

Check if your datetime columns are actually of type datetime, otherwhise convert them before subtracting a timestamp:

In [34]:
all_data['purch_time'] = pd.to_datetime(all_data['purchase_time'])
all_data['visits_time'] = pd.to_datetime(all_data['visit_time'])

In [35]:
all_data['time_to_purchase'] = all_data['purch_time'] - all_data['visits_time']
all_data

user_id           visit_time  \
0     943647ef-3682-4750-a2e1-918ba6f16188  2017-04-07 15:14:00   
1     0c3a3dd0-fb64-4eac-bf84-ba069ce409f2  2017-01-26 14:24:00   
2     6e0b2d60-4027-4d9a-babd-0e7d40859fb1  2017-08-20 08:23:00   
3     6879527e-c5a6-4d14-b2da-50b85212b0ab  2017-11-04 18:15:00   
4     a84327ff-5daa-4ba1-b789-d5b4caf81e96  2017-02-27 11:25:00   
...                                    ...                  ...   
2352  33913ac2-03da-45ae-8fc3-fea39df827c6  2017-03-25 03:29:00   
2353  4f850132-b99d-4623-80e6-6e61d003577e  2017-01-08 09:57:00   
2354  f0830b9b-1f5c-4e74-b63d-3f847cc6ce70  2017-09-07 12:56:00   
2355  b01bffa7-63ba-4cd3-9d93-eb1477c23831  2017-07-20 04:37:00   
2356  0336ca81-8d68-443f-9248-ac0b8ad147d5  2017-11-15 10:11:00   

                cart_time        checkout_time        purchase_time  \
0                     NaN                  NaN                  NaN   
1     2017-01-26 14:44:00  2017-01-26 14:54:00  2017-01-26 15:08:00   
2     2017-08-20 08:31:00                  NaN                  NaN   
3                     NaN                  NaN                  NaN   
4                     NaN                  NaN                  NaN   
...                   ...                  ...                  ...   
2352                  NaN                  NaN                  NaN   
2353                  NaN                  NaN                  NaN   
2354                  NaN                  NaN                  NaN   
2355                  NaN                  NaN                  NaN   
2356                  NaN                  NaN                  NaN   

              purch_time         visits_time time_to_purchase  
0                    NaT 2017-04-07 15:14:00              NaT  
1    2017-01-26 15:08:00 2017-01-26 14:24:00  0 days 00:44:00  
2                    NaT 2017-08-20 08:23:00              NaT  
3                    NaT 2017-11-04 18:15:00              NaT  
4                    NaT 2017-02-27 11:25:00              NaT  
...                  ...                 ...              ...  
2352                 NaT 2017-03-25 03:29:00              NaT  
2353                 NaT 2017-01-08 09:57:00              NaT  
2354                 NaT 2017-09-07 12:56:00              NaT  
2355                 NaT 2017-07-20 04:37:00              NaT  
2356                 NaT 2017-11-15 10:11:00              NaT  

[2357 rows x 8 columns]

In [36]:
all_data.time_to_purchase

0                  NaT
1      0 days 00:44:00
2                  NaT
3                  NaT
4                  NaT
             ...      
2352               NaT
2353               NaT
2354               NaT
2355               NaT
2356               NaT
Name: time_to_purchase, Length: 2357, dtype: timedelta64[ns]

In [37]:
all_data.time_to_purchase.mean()

Timedelta('0 days 00:43:40.954356846')